# Naive Bays Assignment

# Importing Libraries

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

# Loading Dataset

In [2]:
data_train = pd.read_csv("SalaryData_Train.csv")

# Data Understanding

In [3]:
data_train.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [4]:
data_train.shape

(30161, 14)

In [5]:
data_train.isnull().sum()

age              0
workclass        0
education        0
educationno      0
maritalstatus    0
occupation       0
relationship     0
race             0
sex              0
capitalgain      0
capitalloss      0
hoursperweek     0
native           0
Salary           0
dtype: int64

In [6]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30161 entries, 0 to 30160
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   age            30161 non-null  int64 
 1   workclass      30161 non-null  object
 2   education      30161 non-null  object
 3   educationno    30161 non-null  int64 
 4   maritalstatus  30161 non-null  object
 5   occupation     30161 non-null  object
 6   relationship   30161 non-null  object
 7   race           30161 non-null  object
 8   sex            30161 non-null  object
 9   capitalgain    30161 non-null  int64 
 10  capitalloss    30161 non-null  int64 
 11  hoursperweek   30161 non-null  int64 
 12  native         30161 non-null  object
 13  Salary         30161 non-null  object
dtypes: int64(5), object(9)
memory usage: 3.2+ MB


In [7]:
data_train.columns

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

# Converting a categorical data into numerical using Label Encoding

In [8]:
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

In [9]:
LabelEncoder = preprocessing.LabelEncoder()
data_train['workclass'] = LabelEncoder.fit_transform(data_train['workclass'])
data_train['education'] = LabelEncoder.fit_transform(data_train['education'])
data_train['maritalstatus'] = LabelEncoder.fit_transform(data_train['maritalstatus'])
data_train['occupation'] = LabelEncoder.fit_transform(data_train['occupation'])
data_train['relationship'] = LabelEncoder.fit_transform(data_train['relationship'])
data_train['race'] = LabelEncoder.fit_transform(data_train['race'])
data_train['sex'] = LabelEncoder.fit_transform(data_train['sex'])
data_train['native'] = LabelEncoder.fit_transform(data_train['native'])
data_train['Salary'] = LabelEncoder.fit_transform(data_train['Salary'])

In [10]:
data_train

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          5          9           13              4           0   
1       50          4          9           13              2           3   
2       38          2         11            9              0           5   
3       53          2          1            7              2           5   
4       28          2          9           13              2           9   
...    ...        ...        ...          ...            ...         ...   
30156   27          2          7           12              2          12   
30157   40          2         11            9              2           6   
30158   58          2         11            9              6           0   
30159   22          2         11            9              4           0   
30160   52          3         11            9              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 1     4    1         2174            0            40   
1                 0     4    1            0            0            13   
2                 1     4    1            0            0            40   
3                 0     2    1            0            0            40   
4                 5     2    0            0            0            40   
...             ...   ...  ...          ...          ...           ...   
30156             5     4    0            0            0            38   
30157             0     4    1            0            0            40   
30158             4     4    0            0            0            40   
30159             3     4    1            0            0            20   
30160             5     4    0        15024            0            40   

       native  Salary  
0          37       0  
1          37       0  
2          37       0  
3          37       0  
4           4       0  
...       ...     ...  
30156      37       0  
30157      37       1  
30158      37       0  
30159      37       0  
30160      37       1  

[30161 rows x 14 columns]

In [11]:
data_train.describe()

age    workclass    education   educationno  maritalstatus  \
count  30161.000000  30161.00000  30161.00000  30161.000000   30161.000000   
mean      38.438115      2.19933     10.33361     10.121316       2.580087   
std       13.134830      0.95394      3.81226      2.550037       1.498018   
min       17.000000      0.00000      0.00000      1.000000       0.000000   
25%       28.000000      2.00000      9.00000      9.000000       2.000000   
50%       37.000000      2.00000     11.00000     10.000000       2.000000   
75%       47.000000      2.00000     12.00000     13.000000       4.000000   
max       90.000000      6.00000     15.00000     16.000000       6.000000   

         occupation  relationship          race           sex   capitalgain  \
count  30161.000000  30161.000000  30161.000000  30161.000000  30161.000000   
mean       5.959849      1.418322      3.678592      0.675707   1092.044064   
std        4.029633      1.601361      0.834721      0.468118   7406.466611   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      0.000000      4.000000      0.000000      0.000000   
50%        6.000000      1.000000      4.000000      1.000000      0.000000   
75%        9.000000      3.000000      4.000000      1.000000      0.000000   
max       13.000000      5.000000      4.000000      1.000000  99999.000000   

        capitalloss  hoursperweek        native        Salary  
count  30161.000000  30161.000000  30161.000000  30161.000000  
mean      88.302311     40.931269     35.413481      0.248931  
std      404.121321     11.980182      5.956430      0.432401  
min        0.000000      1.000000      0.000000      0.000000  
25%        0.000000     40.000000     37.000000      0.000000  
50%        0.000000     40.000000     37.000000      0.000000  
75%        0.000000     45.000000     37.000000      0.000000  
max     4356.000000     99.000000     39.000000      1.000000

In [12]:
data_train.dtypes

age              int64
workclass        int32
education        int32
educationno      int64
maritalstatus    int32
occupation       int32
relationship     int32
race             int32
sex              int32
capitalgain      int64
capitalloss      int64
hoursperweek     int64
native           int32
Salary           int32
dtype: object

# Model Building

In [13]:
X = data_train.iloc[:,0:13]
y = data_train.iloc[:,13]

In [14]:
X

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          5          9           13              4           0   
1       50          4          9           13              2           3   
2       38          2         11            9              0           5   
3       53          2          1            7              2           5   
4       28          2          9           13              2           9   
...    ...        ...        ...          ...            ...         ...   
30156   27          2          7           12              2          12   
30157   40          2         11            9              2           6   
30158   58          2         11            9              6           0   
30159   22          2         11            9              4           0   
30160   52          3         11            9              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  native  
0                 1     4    1         2174            0            40      37  
1                 0     4    1            0            0            13      37  
2                 1     4    1            0            0            40      37  
3                 0     2    1            0            0            40      37  
4                 5     2    0            0            0            40       4  
...             ...   ...  ...          ...          ...           ...     ...  
30156             5     4    0            0            0            38      37  
30157             0     4    1            0            0            40      37  
30158             4     4    0            0            0            40      37  
30159             3     4    1            0            0            20      37  
30160             5     4    0        15024            0            40      37  

[30161 rows x 13 columns]

In [15]:
y

0        0
1        0
2        0
3        0
4        0
        ..
30156    0
30157    1
30158    0
30159    0
30160    1
Name: Salary, Length: 30161, dtype: int32

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)

In [18]:
X_train.shape, y_train.shape

((24128, 13), (24128,))

In [19]:
X_test.shape, y_test.shape

((6033, 13), (6033,))

# Model Training, Testing and Evaluation

In [20]:
from sklearn.naive_bayes import MultinomialNB
nb_Classifier = MultinomialNB()

In [21]:
nb_Classifier.fit(X_train, y_train)

MultinomialNB()

In [22]:
y_pred = nb_Classifier.predict(X_test)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [23]:
from sklearn.metrics import accuracy_score, precision_score, confusion_matrix, classification_report

In [24]:
accuracy_score(y_test,y_pred)

0.7679429802751533

In [25]:
precision_score(y_test, y_pred)

0.6432865731462926

In [26]:
confusion_matrix(y_test, y_pred)

array([[4312,  178],
       [1222,  321]], dtype=int64)

In [27]:
print("Classification Report:",classification_report(y_test, y_pred))

Classification Report:               precision    recall  f1-score   support

           0       0.78      0.96      0.86      4490
           1       0.64      0.21      0.31      1543

    accuracy                           0.77      6033
   macro avg       0.71      0.58      0.59      6033
weighted avg       0.74      0.77      0.72      6033



# Naive Bayes classification model fro Salary training data give accuracy 76.79%